echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/artifacts/ArtifactMatcherSpec.groovy

```diff
@@ -40,78 +40,255 @@ class ArtifactMatcherSpec extends Specification {
   ]
 
   def constraintsOR = [
-      "one": ["uno", "one"]
+    "one": ["uno", "one"]
   ]
 
   def payloadWithList = [
-      "one": ["one"]
+    "one": ["one"]
   ]
 
   def stringifiedListConstraints = [
     "one": "['uno', 'one']"
   ]
 
+  def jsonPathConstraintsToString = [
+    "\$.one.test1.title": "st"
+  ]
+
+  def jsonPathConstraintsToStringNOT = [
+    "\$.one.test2.title": "no match"
+  ]
+
+  def jsonPathConstraintsToBool = [
+    "\$.one.test1.isValid": "true"
+  ]
+
+  def jsonPathConstraintsToList = [
+    "\$.one.test1.modified": ".yml"
+  ]
+
+  def jsonPathConstraintsToListNOT = [
+    "\$.one.test1.modified": ".xml"
+  ]
+
+  def jsonPathConstraintsToObject = [
+    "\$.one.test1.author": "edgar"
+  ]
+
+  def jsonPathConstraintsToNumber = [
+    "\$.two.test3.count": "3"
+  ]
+
+  def jsonPathConstraintsToListOfObjects = [
+    "\$.two.test2.changes": "bar"
+  ]
+
+  def jsonPathConstraintsToUnknownField = [
+    "\$.two.test4.changes": "bar"
+  ]
+
+  def jsonPathConstraintsBadJsonPath = [
+    "\$.one.test2.changes[a].value": "bar"
+  ]
+
+  def complexPayload = [
+    "one": [
+      "test1": [
+        "title": "test",
+        "modified": ["folder1/file1.txt", "folder1/file2.txt", "folder2/file1.yml"],
+        "isValid": true,
+        "author": [
+          "name": "edgar",
+          "username": "edgarulg"
+        ]
+      ],
+      "test2": [
+        "title": "another test",
+        "created": ["folder3/new.txt"],
+        "isValid": false,
+        "author": [
+          "name": "jorge",
+          "username": "jorge123"
+        ],
+        "changes": [
+          [
+            "id": 1,
+            "value": "foo"
+          ],
+          [
+            "id": 2,
+            "value": "bar"
+          ]
+        ]
+      ]
+    ],
+    "two": [
+      "test3": [
+        "count": 3,
+        "ref": null,
+        "isActive": false
+      ]
+    ]
+  ]
+
   def "matches when constraint is partial word"() {
     when:
-    boolean result = ArtifactMatcher.isConstraintInPayload(shortConstraint, matchPayload)
+    boolean resultOld = ArtifactMatcher.isConstraintInPayload(shortConstraint, matchPayload)
```

Comment:
Is `isConstraintInPayload` being used anywhere else? I don't fully understand why these tests are being combined.

Reply:
Yes, `isConstraintInPayload` is being used in `PubSubHandler`. For this change, I'm only adding the JSONPath feat to the `WebhookHandler`. 

I combined these tests because the `isJsonPathConstraintInPayload` should have the same behavior that `isConstraintInPayload` for backward compatibility.

Reply:
Got it. I might make the variables more descriptive than `resultOld` and `resultNew` (maybe just `isConstraintInPayload` and `isJsonPathConstraintInPayload`). The other thing to consider is just duplicating the tests entirely.